website/static/js/registrationUtils.js

```diff
@@ -0,0 +1,802 @@
+                            }
+                        }
+                    }
+                });
+            }).fail($osf.growl.bind(null, 'Error submitting for review', language.submitForReviewFail));
+        }
+    });
+};
+RegistrationEditor.prototype.submit = function() {
+    var self = this;
+    var currentNode = window.contextVars.node;
+    var currentUser = $osf.currentUser();
+    var messages = self.draft().messages;
+    bootbox.confirm(messages.beforeSubmitForApproval, function(result) {
+        if (result) {
+            $osf.postJSON(self.urls.submit.replace('{draft_pk}', self.draft().pk), {
+                node: currentNode,
+                auth: currentUser
+            }).then(function() {
+                bootbox.dialog({
+                    message: messages.afterSubmitForApproval,
+                    title: 'Pre-Registration Prize Submission',
+                    buttons: {
+                        registrations: {
+                            label: 'Return to registrations page',
+                            className: 'btn-primary pull-right',
+                            callback: function() {
+                                window.location.href = self.draft().urls.registrations;
+                            }
+                        }
+                    }
+                });
+            }).fail($osf.growl.bind(null, 'Error submitting for review', language.submitForReviewFail));
+        }
+    });
+};
+/**
+ * Create a new draft
+ **/
+RegistrationEditor.prototype.create = function(schemaData) {
+    var self = this;
+
+    var metaSchema = self.draft().metaSchema;
+
+    return $osf.postJSON(self.urls.create, {
+        schema_name: metaSchema.name,
+        schema_version: metaSchema.version,
+        schema_data: schemaData
+    }).then(function(response) {
+        var draft = self.draft();
+        draft.pk = response.pk;
+        self.draft(draft);
+    });
+};
+
+/**
+ * Autosave the form state, immediately or deferred.
+ *
+ **/
+RegistrationEditor.prototype.autoSave = $osf.throttle(function () {
+    // TODO(lyndsysimon): Test this.
+    var self = this;
+    // Delay between the completion of a save and the next request
+    var delay = 5000;
+
+    if (!self.isDirty()) {
+        return;
+    }
+
+    var hasBeenSaved = function() {
+        // bool
+        return self.lastSaveRequest() !== undefined;
+    };
+    var delayRemaining = function () {
+        // milliseconds
+        return delay - self.lastSaveAge();
+    };
+    var delaySatisfied = function () {
+        // bool
+        return delayRemaining() <= 0;
+    };
+
+    if(!hasBeenSaved()) {
+        self.save().always(function() {
+            self.autoSave();
+        });
+    } else if (delaySatisfied()) {
+        self.save().always(function() {
+            self.autoSave();
+        });
+    } else {
+        window.setTimeout(self.autoSave.bind(self), delayRemaining());
+    }
+}, 1000);
```

Comment:
Rather than being driven by a user input event (keypress etc), this seems to create an endless loop running in the background. What are the performance implications?